holoviews/core/data.py

```diff
@@ -1308,9 +1334,288 @@ def aggregate(cls, columns, kdims, function, **kwargs):
+            values = columns.data.get(dim.name)
+            return values.flatten() if flat else values
+
+
+    @classmethod
+    def groupby(cls, columns, dim_names, container_type, group_type, **kwargs):
+        # Get dimensions information
+        dimensions = [columns.get_dimension(d) for d in dim_names]
+        kdims = [kdim for kdim in columns.kdims if kdim not in dimensions]
+
+        # Update the kwargs appropriately for Element group types
+        group_kwargs = {}
+        group_type = dict if group_type == 'raw' else group_type
+        if issubclass(group_type, Element):
+            group_kwargs.update(util.get_param_values(columns))
+            group_kwargs['kdims'] = kdims
+        group_kwargs.update(kwargs)
+
+        # Find all the keys along supplied dimensions
+        keys = [columns.data[d.name] for d in dimensions]
+
+        # Iterate over the unique entries applying selection masks
+        grouped_data = []
+        for unique_key in util.cartesian_product(keys):
+            group_data = cls.select(columns, **dict(zip(dim_names, unique_key)))
+            for vdim in columns.vdims:
+                group_data[vdim.name] = np.squeeze(group_data[vdim.name])
+            group_data = group_type(group_data, **group_kwargs)
+            grouped_data.append((tuple(unique_key), group_data))
+
+        if issubclass(container_type, NdMapping):
+            with item_check(False), sorted_context(False):
+                return container_type(grouped_data, kdims=dimensions)
+        else:
+            return container_type(grouped_data)
+
+
+    @classmethod
+    def coord_mask(cls, columns, arr, ind):
+        if isinstance(ind, tuple):
+            ind = slice(*ind)
+        if isinstance(ind, np.ndarray):
+            mask = ind
+        elif isinstance(ind, slice):
+            mask = True
+            if ind.start is not None:
+                mask &= ind.start <= arr
+            if ind.stop is not None:
+                mask &= arr < ind.stop
+        elif isinstance(ind, (set, list)):
+            iter_slcs = []
+            for ik in ind:
+                iter_slcs.append(arr == ik)
+            mask = np.logical_or.reduce(iter_slcs)
+        elif ind is None:
+            mask = None
+        else:
+            index_mask = arr == ind
+            if columns.ndims == 1 and np.sum(index_mask) == 0:
+                data_index = np.argmin(np.abs(arr - ind))
+                mask = np.zeros(len(columns), dtype=np.bool)
+                mask[data_index] = True
+            else:
+                mask = index_mask
+        return mask
+
+
+    @classmethod
+    def select(cls, columns, selection_mask=None, **selection):
+        dimensions = columns.dimensions('key', label=True)
+        val_dims = [vdim for vdim in columns.vdims if vdim in selection]
+        if val_dims:
+            raise IndexError('Cannot slice value dimensions on dense '
+                             'data, convert to sparse format first')
```

Comment:
Similar comment to something I said below...I think 'sparse format' is confusing. As far as I know, you mean 'expanded' format (i.e all kdims have the same shape, no cartesian product applied).